### ISSUE TITLE
[BUG] Getting No Unique API Found error when importing an API into new environment with new version number

### ISSUE THREAD

**Issue by stevebower59** [Axway-API-Management-Plus/apimanager-swagger-promote#218]
**Swagger-Promote version**
1.6.4

**API-Manager and Service-Pack Version**
7.6.2 SP3
Exported a version of API from QA environment, and importing into another Environment.
Target Environment has version v3 of API CerbData as unpublished, and version v4 of API CerbData as published.

I changed the version in api-config.json to v6, and API Name is CerbData, and backend and front-end API paths are the same as the CerbData v3 and v4.  

When I import the API into production, I an ERROR saying App| No Unique API Found.
I tried both importing as state: "unpublished", and state: "published" but get the same error.  I also tried using the -f true flag on the import, but got the same error...

Trace follows:

C:\Axway\swagger-promote-1.6.4\scripts>api-import -s prod -c ..\apis\CerbData-v6\api-config.json -a ..\apis\CerbData-v6\CerbData.json -f true
------------------------------------------------------------------------
API-Manager Promote: 1.6.4 - I M P O R T

To report issues or get help, please visit:
https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote
------------------------------------------------------------------------

0      INFO  nvironmentProperties| Trying to load environment properties from file: env.properties ... not found.
0      INFO  nvironmentProperties| Loaded environment properties from file: env.prod.properties.
3171   INFO     APIManagerAdapter| API-Manager version is: 7.6.2 SP3
3453   INFO  IImportConfigAdapter| Loaded stage API-Config from file: C:\Axway\swagger-promote-1.6.4\apis\CerbData-v6\api-config.prod.json
4000   INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: '..\apis\CerbData-v6\CerbData.json' (relative path)
4312   INFO  IImportConfigAdapter| Handling configured client-applications.
4796   INFO  IImportConfigAdapter| Found existing application: 'CoilData' (null) based on given name 'CoilData'
4796   INFO  IImportConfigAdapter| Found existing application: 'CoilOps' (null) based on given name 'CoilOps'
4796   INFO  IImportConfigAdapter| Found existing application: 'Cerberus' (null) based on given name 'Cerberus'
4812   INFO               Proxies| Sending request to find existing APIs: https://catalog.nov.com:443/api/portal/v1.3/proxies
5265   INFO               Proxies| Found existing API on path: '/CerbData' (published) (ID: '06031aa3-752a-4a38-835e-17a71eeea60a')
5265   INFO               Proxies| Found existing API on path: '/CerbData' (unpublished) (ID: '7c3c8436-7db0-4a40-9e10-b6fc14492c35')
5265   INFO               Proxies| Found: 2 exposed API(s) (apiPath: '/CerbData', filter: null, vhost: 'null', requestedType: proxies)
5265   ERROR                  App| No unique API found
com.axway.apim.lib.AppException: No unique API found
        at com.axway.apim.manager.Proxies.uniqueAPI(Proxies.java:76)
        at com.axway.apim.manager.Proxies.getAPI(Proxies.java:71)
        at com.axway.apim.App.run(App.java:199)
        at com.axway.apim.App.main(App.java:46)

**Expected behavior**
API Is Imported

**Actual behavior**
Error when importing - get Error No Unique API Found

**Log-Output in Debug**

**Test case sample**
> Please describe the steps you have done to replicate the issue

**Comment by stevebower59** [comment]
[CerbData-v6.zip](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3924114/CerbData-v6.zip)

**Comment by cwiechmann** [comment]
Hi @stevebower59, the issue is clear, but I don't understand why it happens. What I can see, is that the API-Path: `/CerbData` instead of `/CerbData/v6` is used and for that path Swagger-Promote finds two APIs. This is a conflict as Swagger-Promote doesn't know which API exactly should be updated.

A bit more background. This is the log line which tells me that the path: `/CerbData` is used to lookup the Actual-API:
`Found: 2 exposed API(s) (apiPath: '/CerbData', filter: null, vhost: 'nu...`
This line is logged by the following code using `requestedApiPath`:
```java
if(foundAPIs.size()!=0) {
	String dbgCrit = "";
	if(foundAPIs.size()>1) 
		dbgCrit = " (apiPath: '"+requestedApiPath+"', filter: "+filter+", vhost: '"+requestedVhost+"', requestedType: "+requestedType+")";
	LOG.info("Found: "+foundAPIs.size()+" exposed API(s)" + dbgCrit);
	return foundAPIs;
}
```
The variable `requestedApiPath` is given to search for the right API. It comes directly from the Desired-API (`desiredAPI.getPath()`).
```java
// Lookup the existing API - If found the actualAPI is valid - desiredAPI is used to control what needs to be loaded
IAPI actualAPI = apimAdapter.getAPIManagerAPI(new Proxies.Builder(APIManagerAdapter.TYPE_FRONT_END).hasApiPath(desiredAPI.getPath()).hasVHost(desiredAPI.getVhost()).build().getAPI(true), desiredAPI);
```
I used the config you shared and on my side it works:
```
0      INFO  nvironmentProperties| Trying to load environment properties from file: env.properties ... not found.
1      INFO  nvironmentProperties| Loaded environment properties from file: env.api-env.properties.
2762   INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: 'CerbData.json' (relative path)
2783   INFO     APIManagerAdapter| API-Manager version is: 7.6.2 SP4
2799   INFO               Proxies| Sending request to find existing APIs: https://api-env:8075/api/portal/v1.3/proxies
2930   INFO               Proxies| Found existing API on path: '/CerbData/v6' (published) (ID: '13c84428-f118-4e9f-bba4-c388584afaea')
2930   INFO               Proxies| Found: 1 exposed API(s)
4440   INFO     APIManagerAdapter| Strategy: Going to update existing API: CerbData (Version: v6)
4441   WARN                   App| No changes detected between Import- and API-Manager-API
```
Do you have any idea why the path `/CerbData` is used on your side?

The API with the given criterias must be unique in order to make sure Swagger-Promote is working on the correct API. The possible criteria are: 
- the API-Path
- the V-Host assigned to that API
- the API-Routing-Key  

The state (published/unpublished) of an API isn't a criteria.  
Please validate in the API-Manager UI that the exposure path used by an API is used only once. 

Just let me know if this makes sense or not. Perhaps you have more details/information you can share.

**Comment by stevebower59** [comment]
The API is configured with a Front-end path of /CerbData/v6  on the lower environment, but when promoted we want to have just /CerbData  front-end Path on Production.   The only difference between the versions is that a different version number is specified in the API metadata  (i.e.  v3, v4, v6) ...   

Since the version isn't one of the distinguishing criteria, I assume that's why the swagger-promote is failing in that case... 

We don't use Api-Routing-Key and use the default VHost in our API configuration.   We just use the API version to distinguish between the different versions that are deployed to the API manager.  

How would we use/configure the API-Routing-Key for our APIs going forward? 

Here is a screenshot of the APIs on our production server that we are trying to promote to. 

![2019-12-05_11-42-47](https://user-images.githubusercontent.com/23459654/70262178-7e7ef880-1759-11ea-87b4-8a560fcb079b.jpg)

**Comment by cwiechmann** [comment]
Okay, that makes sense to me and your assumption is correct. The version you assign to an API is nothing more than just a Meta-Information shown in the Catalog/API-Portal. But not used the API-Manager at runtime to perform API-Matching. And this is where it comes down: At runtime the API-Manager must decide which API should serve the call and this is not possible having a API exposed on the same path.

In order to use the API-Routing-Key you need to turn that feature on. You will find it in the API-Manager settings. However, this setting applies for all APIs. I mean, when turning on, all APIs will have an additional field: 
![image](https://user-images.githubusercontent.com/16292002/70265418-a478bd00-179a-11ea-9b58-b600b620681e.png)
And funny enough, this is considered as a mandatory field in the UI, which I believe is wrong. 

You could do one thing for me: 
When you turn that option on (on a TEST-System :smirk:), will your APIs still being served by the API-Manager even those APIs doesn't have a Routing-Version number? If yes, I would say it is a bug in the UI, which I have unfortunately taken over into Swagger-Promote. 
Meaning, also Swagger-Promote rejects to register an API without an API-Routing-Key when that option is turned on in the API-Manager. However, I'm happy to remove that check, if APIs are also working without that Routing-Key filled. Which is obviously the case for all of your APIs before you turned that option on. 

I would understand that Version as to be optional: 
If an API is registered without that version, that API should be used by all Clients **not providing** the version number. Like a default routing. 
If an API is registered with that Routing-Version and Clients **present the correct version**, that API is selected by the API-Manager.
Can you perhaps check that? Actually no time on my side to do that.

**Comment by cwiechmann** [comment]
Hi @stevebower59, following that discussion I have checked how the API-Manager behaves in terms of the Query-Based-Routing. For more details please see here: #222 
With that change, which will become available in version 1.6.5, you can make use of Query-Based-Routing to register the same API multiple times, giving each a different Routing-Version besides one, which behaves like a default.
